### DISCUSSION TITLE
Nested , collapsing and expansive lists

### DISCUSSION THREAD

**Discussion by verdaux** [GitJournal/GitJournal#783] [Feature Requests]
Like I suggested on playstore, a workflowy like nested list option would be great addition

---
> Please vote on this issue with a 👍 reaction. It's one of the primary mechanisms used to prioritize issues. Please also subscribe to the issue if you're interested, as that way I can discuss possible implementations and share mockups.

**Comment by vHanda** [comment]
Hey Carter. The UI part I understand, I can make this into a different editor. Maybe even call it "List Editor", unless you have a better name.

I'm not too sure how to map this into files. The simplest would be that it just maps to one file. Clicking on a list item, would hide the rest and just enlarge that item.

Or

Every bullet point count map to its own file?

**Comment by tsbertalan** [comment]
Well, regular github-flavored markdown does allow for nested checklists:

```markdown
- [ ] a
    - [ ] b
```

gives

- [ ] a
    - [ ] b

Could syntax this be supported in the app? At the moment, if you edit the files elsewhere to have such indentation, they will render as nested lists in the card and grid views, but not in the existing list editor. (However, rearranging nested tasks in that editor results in the corresponding lines in the makdown file being rearranged, without losing the indentation.)

**Comment by tsbertalan** [comment]
Ah, I see that #161 is already about this. In my limited tests, workflowy seems to be a tree editor, with the obvious UI feature of collapsing subtrees. The slightly less-obvious feature is that, in their UI, collapsed subtrees become links to an uncollapsed view rooted at that subtree. This is an elegant UI solution for navigating trees of (apparently) unlimited depth, but I'd think it would violate the convention that seems to be in place for GitJournal of having one view per markdown file.

**Comment by vHanda** [comment]
We could still have some way of zooming into a subtree. We would need to experiment on how to express that from a UI point of view, but it could be done.

Then you would basically just keep your notes in one large file. Another user also requested an option to directly open a note on startup instead of the default view. Those 2 things combined could make it work for you.

Though everything is a checklist, is that fine?

**Comment by verdaux** [comment]
> Ah, I see that #161 is already about this. In my limited tests, workflow seems to be a tree editor, with the obvious UI feature of collapsing subtrees. The slightly less-obvious feature is that, in their UI, collapsed subtrees become links to an uncollapsed view rooted at that subtree. This is an elegant UI solution for navigating trees of (apparently) unlimited depth, but I'd think it would violate the convention that seems to be in place for GitJournal of having one view per markdown file.

I see your point. Looks like worth a shot.

**Comment by verdaux** [comment]
> We could still have some way of zooming into a subtree. We would need to experiment on how to express that from a UI point of view, but it could be done.
> 
> Then you would basically just keep your notes in one large file. Another user also requested an option to directly open a note on startup instead of the default view. Those 2 things combined could make it work for you.
> 
> Though everything is a checklist, is that fine?

My intention was to inculcate the outlining note-taking method that really helps me memorize points faster. The UI question can be solved by keeping a separate section for just these types of files. Just an idea. I don't mean to twist the beauty of your app by introducing this. I love using it truly. What are your thoughts?

**Comment by vHanda** [comment]
> > We could still have some way of zooming into a subtree. We would need to experiment on how to express that from a UI point of view, but it could be done.
> > Then you would basically just keep your notes in one large file. Another user also requested an option to directly open a note on startup instead of the default view. Those 2 things combined could make it work for you.
> > Though everything is a checklist, is that fine?
> 
> My intention was to inculcate the outlining note-taking method that really helps me memorize points faster. The UI question can be solved by keeping a separate section for just these types of files. Just an idea. I don't mean to twist the beauty of your app by introducing this. I love using it truly. What are your thoughts?

I lean towards utility and the app being useful to people. My main priority is making it easier for people to be in control of their data. So I don't have strong opinions on the different UIs. Each person is different and no one method works for everyone.

I've used workflowy before and quite liked it. I can try to experiment with recreating such an experience. It could be a new Editor separate from the Checklist editor.

**Comment by verdaux** [comment]
Your app is most definitely useful. I can vouch for that. And has helped me immensely. Thanks for putting it out in the universe. More power to you. 

Let me know if we can experiment with the separate editor.

**Comment by tsbertalan** [comment]
@vHanda 

> keep your notes in one large file

It's worth noting that it doesn't have to be either/or--you could have such a zooming editor mode that allows for viewing subtrees rooted only within the bounds of one .md file. Or even over a whole file tree, though it should probably be made visually distinct whether two nodes are siblings due to being .md files in the same folder, or child/parent due to folder structure (If you want to include folder hierarchy in the tree structure, that is).

An issue with trying to fit *all* of your chosen markdown subset into a Workflowy-like UI would be deciding what to do with non-checklist items. Where do headings, code fragments, images, quotations, etc. etc. go in the tree? I've been thinking and prototyping on this for my own use for the past month or so, and I'm now inclined to follow the semantic structure of headings/subheading in the tree, and to follow indentation (fairly) strictly for determining childhood. E.g., in 

```markdown
- One

    - [x] Two

        Three

    Four

        Five
````

(GitHub renders that as:

- One

    - [x] Two

        Three

    Four

        Five
    

)

`Three` (preceded by 4=2*k spaces) is a child of `Two`, but `Four` is a child of `One`. (Note that I put extra newlines between these because that seems to be required between elements by e.g. Typora in its "strict" mode.) It seems to be common to *not* subsequently treat `Five` as a child of `Four`, but instead as a code block under `One`, which seems to be a break from the nice rule. However, that doesn't mean that this markdown->semantic structure conversion wouldn't still be predictable and deterministic.

> Though everything is a checklist, is that fine?

My thoughts above about how to include non-checklist items above are related to this, but another point I'd like to make is that almost all outlining programs, in my experience, are focused on, well, outlines. There are very few (in fact, [Leo](leoeditor.com) is the only one I think I've encountered) that allow for full DAGs instead of just trees (or maybe forests, if you count different .md files not as being part of a unified overarching structure). I believe that DAGs are fundamentally a better fit for describing the dependency hierarchy that is a tasklist--two tasks could both share a common blocker/prerequisite task.

And, for note-taking (@verdaux 's apparent use-case), there's no reason the graph should even be acyclic.

I don't think there is a perfect way to do this in an all-in-one flat-file representation--you need to choose one instance as the master, and the other needs to refer to it. Leo (which stores data in XML) handles this by having all items be ID-references to data stored in a separate section of the document. One way this can be emulated in markdown is by having the references be links (either the `[label](dest)` kind, or the [Obsidian](obsidian.md)/[Zettler](zettlr.com)/otherwise `[[wikilink]]` kind) to separate markdown files (or at least to separate `#`'d headings within the same document), which are considered by the system to be transcluded-in-place-with-indentation.

I think this is likely beyond the scope of what you want to do with this program, but I thought I'd share in case it gripped you strongly.

**Comment by verdaux** [comment]
@tsbertalan Thank you for your work and findings on this. As you pointed out a nested checklist is all I ask. I do not wish to include other object types such as images or different data styles. 

My only vision is to have a single purpose nested checklist markdown document whose indentation indicates a sub-point sort of representation.

It could be bullet points instead of checklist or any other symbol just as long as my viewing of them signifies a nested theme or hierarchy.

For example:

   * Country
     * State
       * City
         * District
           * Street

I know I can manage this sort of indentation with added spaces characters in my regular document. But this feature in the checklist shall save me that task.


         

@vHanda

**Comment by akvadrako** [comment]
I think even better than nested lists would be support for headings in the Checklist editor. Currently I have for example a `pack.md` file with sections for different situations. It would be nice if when I view it I could see larger, tickable boxes.

Of course both nested lists and sections would be better.

**Comment by vHanda** [comment]
> I think even better than nested lists would be support for headings in the Checklist editor. Currently I have for example a `pack.md` file with sections for different situations. It would be nice if when I view it I could see larger, tickable boxes.
> 
> Of course both nested lists and sections would be better.

That's actually much easier to implement. There is another bug tracking it. Could be please vote for it? https://github.com/GitJournal/GitJournal/issues/96
